From 91a0ddd8f13f02a1250ba9e15db72be0f1fc7c9e Mon Sep 17 00:00:00 2001
From: Xin Jiang <jiangxin@hygon.cn>
Date: Thu, 13 Jul 2023 09:35:10 +0800
Subject: [PATCH 9/9] anolis: vga: force full update for CSV guest

As CSV's NPT(nested page table) is managed by firmware, VMM is hard
to track the dirty pages of vga buffer. Although VMM could perform
a command to firmware to update read/write attribute of vga buffer
in NPT, it costs more time due to communication between VMM and
firmware. So the simplest method is to fully update vga buffer
always.

Signed-off-by: Xin Jiang <jiangxin@hygon.cn>
Change-Id: I093cebad952157aa38edf43fcf861926f7204dd6
---
 hw/display/vga.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/hw/display/vga.c b/hw/display/vga.c
index 0cb26a791b..1b5bd5f363 100644
--- a/hw/display/vga.c
+++ b/hw/display/vga.c
@@ -37,6 +37,9 @@
 #include "migration/vmstate.h"
 #include "trace.h"
 
+#include "target/i386/sev.h"
+#include "target/i386/csv.h"
+
 //#define DEBUG_VGA_MEM
 //#define DEBUG_VGA_REG
 
@@ -1781,6 +1784,11 @@ static void vga_update_display(void *opaque)
             s->cursor_blink_time = qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL);
             full_update = 1;
         }
+
+        /* Force to full update in CSV guest. */
+        if (csv_enabled())
+            full_update = 1;
+
         switch(graphic_mode) {
         case GMODE_TEXT:
             vga_draw_text(s, full_update);
-- 
2.17.1

